### PR TITLE
Improve hstack lowering

### PR DESCRIPTION
**Pull request by rjzamora**
## Description
- This is a follow-up to https://github.com/rapidsai/cudf/pull/21796
- This (hopefully) simplifies some code in https://github.com/rapidsai/cudf/pull/22191

**Problem statement**: We currently translate `HStack` nodes with non-pointwise expressions to the equivalent `Select` node at lowering time. This is because all our non-pointwise `Expr`-decomposition logic is specific to `Select`. Before this PR, this translation was skipped whenever the underlying `HStack` was completely overwriting it's original columns. The problem with this case is that we loose "anchor" columns that tell the `Select` how to broadcast scalar-aggregation results.

**Proposed solution**: We add a temporary "anchor" column to the translated `HStack` so that broadcasting works correctly in the `Select` node.

**Motivation**:
- We can handle all `over()` expression decomposition within `Select` if we know **all** non-pointwise HStack operations are lowered to `Select` anyway.
- We don't "fall back" for other non-`over` `HStack` corner cases either.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf_polars/cudf_polars/experimental/parallel.py [hunk 1]
```diff
@@ -460,27 +463,51 @@ def _(
...
-                rec,
-            )
-        # All output columns are aggregations: no N-row passthrough to anchor
-        # broadcast. Fall back so HStack.do_evaluate uses target_length=child.num_rows.
-        return _lower_ir_fallback(ir, rec)
+        if ir.should_broadcast and all(name in col_map for name in ir.schema):
+            # We need to add a temporary anchor column to preserve row count.
+            child, anchor_name, anchor_dtype = _add_anchor_column(ir)
+            new_schema = ir.schema | {anchor_name: anchor_dtype}
+            ir = HStack(new_schema, ir.columns, ir.should_broadcast, child)
```
Comment (Matt711) [inline python/cudf_polars/cudf_polars/experimental/parallel.py]: nitpick

Seems like you can just build the new schema without making an `HStack`

**Review by Matt711 — APPROVED** [review]
Looks good one nitpick. And I think we can write a test here.

File: python/cudf_polars/cudf_polars/experimental/parallel.py (hunk 1 above)
Comment (rjzamora) [reply python/cudf_polars/cudf_polars/experimental/parallel.py]: Not sure what you mean?

**Comment by rjzamora** [comment]
>I think we can write a test here.

So, I'm not entirely sure if a new test makes sense. `test_with_columns_scalar_upstream_20981` would fail if this code path was implemented incorrectly. The only reason we didn't get an error before is that we didn't define a `msg` argument to the `_lower_ir_fallback` call (so we didn't see a `UserWarning` failure here). Now that there is no silent fallback path, it needs to "work" for the test to pass.

File: python/cudf_polars/cudf_polars/experimental/parallel.py (hunk 1 above)
Comment (rjzamora) [reply python/cudf_polars/cudf_polars/experimental/parallel.py]: **Thinking out loud**: We use `_add_anchor_column` to tack on a temporary int8 column before our problematic `HStack` operation. We must do this before (and separate from) the problematic `HStack` so that we know the temporary column will be broadcasted to the correct length. After that, we must update the original (problematic) `HStack` to use the new schema and child. That last part is what this line is doing.

Comment (Matt711) [reply python/cudf_polars/cudf_polars/experimental/parallel.py]: something like this?
```diff
         child: IR = ir.children[0]
         anchor_name: str | None = None
         col_map = {ne.name: ne for ne in ir.columns}
+        schema = ir.schema
         if ir.should_broadcast and all(name in col_map for name in ir.schema):
             # We need to add a temporary anchor column to preserve row count.
             child, anchor_name, anchor_dtype = _add_anchor_column(ir)
-            new_schema = ir.schema | {anchor_name: anchor_dtype}
-            ir = HStack(new_schema, ir.columns, ir.should_broadcast, child)
+            schema = ir.schema | {anchor_name: anchor_dtype}
         exprs = tuple(
             col_map[name] if name in col_map else NamedExpr(name, Col(dtype, name))
-            for name, dtype in ir.schema.items()
+            for name, dtype in schema.items()
         )
-        new_ir: Select | Projection = Select(
-            ir.schema, exprs, ir.should_broadcast, child
-        )
+        new_ir: Select | Projection = Select(
+            schema, exprs, ir.should_broadcast, child
+        )
```
